### ISSUE TITLE
Bring pulse command naming in line with circuit Registers

### ISSUE THREAD

**Issue by taalexander** [Qiskit/qiskit-terra#2593]
### What is the expected enhancement?
If a `Command` does not have a `name` it is currently assigned a default name of `p%d` where `%d` is the number of existing `Command`s. Instead the naming commands should be brought in line with the `circuit.register.Register` classes default naming policy. There should be a unique prefix and counter for each `Command` type.

Additionally, validation of the name type should also be performed in a similar manner to the register.

Suggestion for prefix names:
- `Command` default: `c`
- `SamplePulse`: `p`
- `FrameChange`: `fc`
-  `PersistentValue`: `pv`
- `SamplePulse`: `acq`
- `Snapshot`: `snap`

**Comment by eggerdj** [comment]
@taalexander I agree with the naming convention. I think this will help make the QObj more readable. Can you assign this to me? I propose creating a small function `create_name(key: str)` in class `Command` which could be called from the constructors of the SamplePulse, FrameChange, etc... e.g. `__init__` in FrameChange would call `create_name('acq')` to set `self.name`. What type of user given names should we accept? Anything that can be cast to a str?

**Comment by taalexander** [comment]
Thank you for taking this @eggerdj. I believe 'Register' accepts names of the form. '[a-z][a-zA-Z0-9_]*'. Perhaps we should also take the time to make `Register` use `create_name` in a following PR so as to unify the two naming schemes.

**Comment by eggerdj** [comment]
One function for both Register and PulseCommand and its children makes sense. Where should we put the function?

**Comment by eggerdj** [comment]
Here is what I had in mind for `create_name`:

```
def create_name(name: str = None, prefix: str = None, counter: itertools.count = None) -> str:
    if name is None:
        try:
            name = '%s%i' % (prefix, next(counter))
        except TypeError:
            raise QiskitError("prefix and counter must be non-None when name is None.")
    else:
        try:
            name = str(name)
        except Exception:
            raise QiskitError("The circuit/pulse command name should be castable to a string "
                              "(or None for autogenerate a name).")
        name_format = re.compile('[a-z][a-zA-Z0-9_]*')
        if name_format.match(name) is None:
            raise QiskitError("%s is an invalid OPENQASM register/OpenPulse command name." % name)
            
    return name
```

But I'm not sure where it should go. If it would not be used by `class Register` I would have it live in 'class Command'.

**Comment by taalexander** [comment]
I agree it should live as a `classmethod` in the `Command` class for now.  Too build the custom counter for each subclass is a bit more difficult. I believe a metaclass will be necessary, eg., [usecases for metaclasses](https://stackoverflow.com/questions/392160/what-are-some-concrete-use-cases-for-metaclasses) as `init_subclass` is Python 3.6+. If this is too complicated I am ok with only having a single counter or all commands for now.

**Comment by taalexander** [comment]
Closed by #2623.
